### PR TITLE
Grains approaches

### PR DESCRIPTION
**Pull request by bobahop**
Initial commits for file structure and code. Explanatory details to follow...

### CODE REVIEW

**Comment by bobahop** [comment]
@ErikSchierboom Grains approaches  is ready for review.

File: exercises/practice/grains/.approaches/bit-shifting/content.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+public static class Grains
+{
+    public static ulong Square(int n)
+    {
+        switch (n)
+        {
+            case int num when num > 0 && num < 65: return (ulong)1 << (num - 1);
+            default: throw new ArgumentOutOfRangeException("n must be 1 through 64");
+        }
```
Comment (ErikSchierboom) [inline exercises/practice/grains/.approaches/bit-shifting/content.md:15]: While I like the switch, I find the below easier to read:

```suggestion
        if (num < 1 || num > 64)
            throw new ArgumentOutOfRangeException("n must be 1 through 64");

        return (ulong)1 << (num - 1);
```

Interestingly, you use this structure in the other approach. Was there any particular reason why?

BTW I'm not opposed to having a separate section on how a switch can also be used, as it is a nice approach.

File: exercises/practice/grains/.approaches/bit-shifting/snippet.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+    public static ulong Square(int n)
+    {
+        switch (n)
+        {
+            case int num when num > 0 && num < 65: return (ulong)1 << (num - 1);
+            default: throw new ArgumentOutOfRangeException("n must be 1 through 64");
+        }
+    }
```
Comment (ErikSchierboom) [inline exercises/practice/grains/.approaches/bit-shifting/snippet.txt:8]:
```suggestion
public static ulong Square(int n)
{
    switch (n)
    {
        case int num when num > 0 && num < 65: return (ulong)1 << (num - 1);
        default: throw new ArgumentOutOfRangeException("n must be 1 through 64");
    }
}
```

File: exercises/practice/grains/.approaches/max-value/snippet.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+public static ulong Total()
+{
+  return System.UInt64.MaxValue;
```
Comment (ErikSchierboom) [inline exercises/practice/grains/.approaches/max-value/snippet.txt:3]:
```suggestion
    return System.UInt64.MaxValue;
```

File: exercises/practice/grains/.approaches/pow/snippet.txt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+    public static double Square(int i)
+    {
+        if (i is <= 0 or > 64) throw new ArgumentOutOfRangeException(nameof(i));
+
+        return Math.Pow(2, i - 1);
+    }
+    
+    public static double Total() => Enumerable.Range(1, 64).Sum(Square);
+
```
Comment (ErikSchierboom) [inline exercises/practice/grains/.approaches/pow/snippet.txt:9]:
```suggestion
public static double Square(int i)
{
    if (i is <= 0 or > 64) throw new ArgumentOutOfRangeException(nameof(i));

    return Math.Pow(2, i - 1);
}

public static double Total() => Enumerable.Range(1, 64).Sum(Square);
```

File: exercises/practice/grains/.articles/performance/code/Program.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Running;
+using System.Linq;
+using System.Numerics;
+
+[MemoryDiagnoser]
+public class BenchMe
+{
+    // private readonly string word = "Thumbscrew-Japingly";
```
Comment (ErikSchierboom) [inline exercises/practice/grains/.articles/performance/code/Program.cs:9]:
```suggestion
```

**Review by ErikSchierboom — APPROVED** [review]

File: exercises/practice/grains/.approaches/bit-shifting/content.md (hunk 1 above)
Comment (bobahop) [reply exercises/practice/grains/.approaches/bit-shifting/content.md:15]: No particular reason. It's just how I rolled for this one, and it's a nice variety from the other approach. I'll leave it for now. How a switch works is pretty basic, and I link to the doc, so I'll not add a section for now.

File: exercises/practice/grains/.approaches/max-value/snippet.txt (hunk 3 above)
Comment (bobahop) [reply exercises/practice/grains/.approaches/max-value/snippet.txt:3]: It kills me that sometimes I get two-space indentation and sometimes four. Haven't kept track of where they differ.

**Review by ErikSchierboom — APPROVED** [review]

**Pull request merged**